File: core/streaming/live_update_engine.py
from __future__ import annotations

from typing import Any, Dict, List

from core.streaming.stream_capture_engine import make_stream_event

MAX_UPDATES = 5000


def track_live_runtime_updates(page: Any) -> Dict[str, Any]:
    updates: List[Dict[str, Any]] = []
    events: List[Dict[str, Any]] = []

    if page is not None and hasattr(page, "_test_live_updates"):
        updates = list(page._test_live_updates)[:MAX_UPDATES]

    for index, update in enumerate(updates):
        events.append(
            make_stream_event(
                step=index,
                source="live_update",
                direction=str(update.get("kind", "refresh")),
                payload=str(update.get("payload", "")),
                connection_id=str(update.get("target", "")),
            )
        )

    return {
        "updates": updates,
        "events": events,
        "bounded": True,
    }

File: src/streaming/liveUpdateEngine.ts
/**
 * Converted from Python: core/streaming/live_update_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { makeStreamEvent } from "./streamCaptureEngine.js";

export let MAX_UPDATES: any = 5000;
export function trackLiveRuntimeUpdates(page: any): any {
  var updates: any[] = [];
  var events: any[] = [];
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("_test_live_updates") in (page as object) || typeof (page as Record<string, unknown>)[String("_test_live_updates")] === "function")))) {
    updates = py.slice([...py.iter(page._test_live_updates)], null, MAX_UPDATES);
  }
  var index: any;
  var update: any;
  for ([index, update] of py.enumerate(updates)) {
    py.listAppend(events, makeStreamEvent(index, "live_update", py.toStr(py.get(update, "kind", "refresh")), py.toStr(py.get(update, "payload", "")), py.toStr(py.get(update, "target", ""))));
  }
  return {"updates": updates, "events": events, "bounded": true};
}
export { makeStreamEvent };
